Add tests

aljones15 opened the pull request:
Adds basic tests for sessionService and session:

```js
  sessionService API
    unauthenticated request
      ��� should get a session with no account
    authenticated request
      ��� should get a session with an account
      ��� should logout a session
      ��� should expire after 1 second
      ��� should refresh on get
  session API
    unauthenticated request
      ��� should create a session
      ��� should get a session with no data
    authenticated request
      ��� should get a session with data
      ��� should logout a session
      ��� should expire after 1 second
      ��� should refresh
      ��� should emit change event on refesh when expired
      ��� should emit change event on end
      ��� should emit change event if authentication passed to refresh

Finished in 12.417 secs / 8.727 secs @ 18:25:26 GMT-0400 (Eastern Daylight Time)

SUMMARY:
��� 14 tests completed
TOTAL: 14 SUCCESS
All tests passed.
```

aljones15 committed:
Add deps.

aljones15 committed:
Add an unauthenticated test.

aljones15 committed:
Add authenticated session test.

aljones15 committed:
Add a logout test.

aljones15 committed:
Remove get API test suite.

aljones15 committed:
Add express session settings.

aljones15 committed:
Make test sessions expire after 1 second.

aljones15 committed:
Add tests for refresh.

aljones15 committed:
Add session tests.

aljones15 committed:
Add createSession test.

aljones15 committed:
Add more assertions on session.data.

aljones15 committed:
Add change event tests.

aljones15 committed:
Add copyrights to session tests and helpers.

aljones15 committed:
Remove redudant comments.

mattcollier commented on test/test.config.js:
```diff
@@ -24,3 +26,29 @@ config.mongodb.dropCollections.collections = [];
+config.express.session.resave = true;
+config.express.session.saveUninitialized = false;
+
+// FIXME change this to a common setting
```
What does this mean?  Can it be resolved?

aljones15 replied:
this is being resolved in the session expiration work that this test suite will help with. So basically in the PR for session expire this FIXME should be removed. This is just a reminder to do it then.

mattcollier replied:
Please remove this then as it's not actionable by anyone but yourself and create a github issue if you need a reminder to do something.  Generally speaking, adding TODO's FIXME's to code going forward is not really going to be a thing.  It's too difficult to identify and triage FIXMEs and TODOs months after the fact.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/68a177b0cbcb2b52ac7eb1f27fe6acbb50cf8f26

Issue added: https://github.com/digitalbazaar/bedrock-web-session/issues/10

mattcollier commented on test/web/10-sessionService.js:
```diff
@@ -0,0 +1,138 @@
+      should.not.exist(err);
+      should.exist(session);
+      session.should.be.an('object');
+      const keys = Object.keys(session);
+      // an unauthenticated session has no data
+      keys.should.deep.equal([]);
```
```suggestion
      // an unauthenticated session has no data
      session.should.eql({});
```
`eql` is a shorthand for `deep.equal` 

Please use `foo.should.have.keys(['a', 'b']);` for this sort of assertion elsewhere in this test suite.

mattcollier commented on test/web/10-sessionService.js:
```diff
@@ -0,0 +1,138 @@
+    });
+  }); // end unauthenticated request
+  describe('authenticated request', () => {
+    let session, account, totp, email, password = null;
```
code stye does not allow defining multiple variables on one line, please separate each def to a separate line.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/94f6ce22182e0830e69a81e9f20db591ea4337e4

mattcollier commented on test/web/10-sessionService.js:
```diff
@@ -0,0 +1,138 @@
+    it('should get a session with an account', async () => {
+      let err;
+      try {
+        session = await sessionService.get();
```
is the code simpler if `session` is scoped to each `it`?  If one `it` is reusing `session` from a prior `it` that is a pattern to be avoided if possible.

aljones15 replied:
That's a really good idea for this test so I did that: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/68a177b0cbcb2b52ac7eb1f27fe6acbb50cf8f26

mattcollier commented on test/web/10-sessionService.js:
```diff
@@ -0,0 +1,138 @@
+      session.account.should.be.an('object');
+      session.account.should.have.property('id');
+      session.account.id.should.equal(account.id);
+      for(let i = 0; i < 4; i++) {
```
I think 5 iterations here is more to the point.  Also, please add a comment that you are demonstrating that the session would have expired at 1000ms if the refresh were not working properly.  Just provide the reader more information about the aims of the test and how your are accomplishing them.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/57c346b2c53b1f3110f15b5356fc0f937a0cbaec

mattcollier commented on test/web/20-session.js:
```diff
@@ -0,0 +1,270 @@
+      session.data.account.should.be.an('object');
+      session.data.account.should.have.property('id');
+      session.data.account.id.should.equal(account.id);
+      for(let i = 0; i < 4; i++) {
```
same comment as above.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/57c346b2c53b1f3110f15b5356fc0f937a0cbaec

mattcollier commented on test/web/20-session.js:
```diff
@@ -0,0 +1,270 @@
+        session.data.account.id.should.equal(account.id);
+        const changeEvent = new Promise(resolve => {
+          session.on('change', ({authentication, oldData, newData}) => {
+            should.not.exist(authentication);
+            should.exist(oldData);
+            should.exist(newData);
+            oldData.should.not.deep.equal(newData);
```
if any of these things throw (the assertion fails), then it should call `reject(err);` on your Promise.  So, all these assertions should be in a try/catch, catch the error and call `reject` with it.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/886ccc59a3e6990576c161773af7940f24466e53

mattcollier commented on test/web/20-session.js:
```diff
@@ -0,0 +1,270 @@
+      session.data.account.should.have.property('id');
+      session.data.account.id.should.equal(account.id);
+      const changeEvent = new Promise(resolve => {
+        session.on('change', ({authentication, oldData, newData}) => {
```
same comment as above.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/886ccc59a3e6990576c161773af7940f24466e53

mattcollier commented on test/web/20-session.js:
```diff
@@ -0,0 +1,270 @@
+        keys.should.deep.equal(['data', '_service', '_eventTypeListeners']);
+        Object.keys(session.data).should.deep.equal(['account']);
+        await changeEvent;
+        authSpy.withArgs({
```
nice!

mattcollier commented on test/web/helpers.js:
```diff
@@ -0,0 +1,58 @@
+  const exists = await accountService.exists({email});
+  if(exists) {
+    // if the account exists just return
+    return;
```
I think the better pattern here is to throw a `DuplicateError` and write the caller to handle `DuplicateError`.

This does beg the question, what does `accountService.create` do already?  Maybe this existence check is not needed if `create` already throws a `DuplicateError`.

aljones15 replied:
this is for setting up test data, however it does mean the failure case is not throwing so I will add a `DuplicateError`.

aljones15 replied:
Addressed this here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/4e251232ae3d11ad00b1a4a4e74499e85c8084d0

mattcollier commented on test/web/helpers.js:
```diff
@@ -0,0 +1,58 @@
+    authenticationMethod: 'password-test',
+    serviceId: short_name
+  });
+  return {account, email, password, totp};
```
returning values that have just been passed into this function: `email`, `password` is an anti-pattern.  Please only return new data that is generated by this function.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/94f6ce22182e0830e69a81e9f20db591ea4337e4

mattcollier reviewed (changes requested):
Thanks for working on this.  This is an important feature that we've needed for a long-long time.

aljones15 committed:
Use eql shorthand for deep.equal.

Co-authored-by: mattcollier <902041+mattcollier@users.noreply.github.com>

aljones15 committed:
Scope session to each test.

aljones15 committed:
Add comments to explain refresh tests.

  - Use eql over deep.equal

aljones15 committed:
createAccount only returns account & totp.

aljones15 committed:
Add try catch to promises to ensure no unhandled rejections occur in tests.

aljones15 committed:
Ensure logout before account creation.

aljones15 committed:
Use should.have.keys over Object.keys

  - Reuse the same session id for session.js tests
  - Clean up session store after tests

aljones15 committed:
Throw if duplicate account in createAccount.

aljones15 committed:
Add a delay to reduce CI flakiness.

aljones15 committed:
Run tests in new windows not iFrames.

aljones15 committed:
Try with concurrency set to 1.

aljones15 committed:
Try using iframe again.

aljones15 committed:
Remove new window code.

dlongley commented on test/test.config.js:
```diff
@@ -24,3 +29,28 @@ config.mongodb.dropCollections.collections = [];
+config.express.session.prefix = 'web-authn-token-test';
+// test specific settings
+config.express.session.rolling = true;
+config.express.session.resave = true;
```
I don't quite understand what's going on here. Are these the defaults? If not, why are they being set here -- don't we want to (also/at least) test against the defaults? If they are the defaults, why are they being set here?

aljones15 replied:
addressed here: https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/878d7d92c8856c6ea08ec0500d428bca88096309

aljones15 committed:
Use default session values.

aljones15 committed:
Update test deps.

aljones15 committed:
Upgrade bedrock-authn-token too.

aljones15 committed:
Try 1 second wait to ensure ttl has finished before starting next auth test.

aljones15 commented:
I ran the last [commit](https://github.com/digitalbazaar/bedrock-web-session/pull/9/commits/812baca1c7338f8a3d49c614d3b9e1914dab6d80) 3 times via github actions and did not get the timing error many previous commits got.

mattcollier reviewed (approved):
TY.  Merging.
